File: packages/libARUtils/TestBench/iOS/TestBLEFtp/TestBLEFtp/autoTest.m
/*
    Copyright (C) 2014 Parrot SA

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in
      the documentation and/or other materials provided with the 
      distribution.
    * Neither the name of Parrot nor the names
      of its contributors may be used to endorse or promote products
      derived from this software without specific prior written
      permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
    OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
    AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.
*/
//
//  autoTest.m
//  ARTest
//
//  Created by David on 13/05/2014.
//  Copyright (c) 2014 ___PARROT___. All rights reserved.
//

#import <libARSAL/ARSAL_CentralManager.h>

#import <libARDiscovery/ARDISCOVERY_BonjourDiscovery.h>
//#import "ARNETWORKAL_BLEManager.h"
//#import "ARNETWORKAL_BLENetwork.h"
//#import "ARNETWORKAL_Manager.h"
#include <libARSAL/ARSAL_Error.h>
#include <libARSAL/ARSAL_Singleton.h>
#include <libARSAL/ARSAL_BLEManager.h>
#import "ARUTILS_BLEFtp.h"

#include <libARSAL/ARSAL_MD5_Manager.h>

#import "autoTest.h"



@implementation autoTest

- (void)testMd5
{
    eARSAL_ERROR result = ARSAL_OK;
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    const char* doc = [[paths lastObject] cString];
    char path[512];
    uint8_t md5[16];
    //char md5Txt[33];
    
    ARSAL_MD5_Manager_t *manager = ARSAL_MD5_Manager_New(&result);
    
    ARSAL_MD5_Manager_Init(manager);
    
    strcpy(path, doc);
    strcat(path, "/txt.txt");
    
    FILE *file = fopen(path, "wb");
    fwrite("123\n", 1, 4, file);
    fflush(file);
    fclose(file);
    
    result = ARSAL_MD5_Manager_Compute(manager, path, md5, sizeof(md5));
    
    result = ARSAL_MD5_Manager_Check(manager, path, "ba1f2511fc30423bdbb183fe33f3dd0f");
    
    ARSAL_MD5_Manager_Close(manager);
    
    ARSAL_MD5_Manager_Delete(&manager);
    
}

- (void)testConnection
{
    //[self performSelectorInBackground:@selector(testScann) withObject:nil];
    
    NSThread *thread = [[NSThread alloc] initWithTarget:self selector:@selector(testScann) object:nil];
    [thread start];
    
    //[self testMd5];
}

- (void)testScann
{
    ARBLEService *foundService = nil;
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    //const char* doc = [[paths lastObject] cString];
    NSString *doc = [paths lastObject];
    
    [[ARDiscovery sharedInstance] start];
    
    while (foundService == nil)
    {
        [NSThread sleepForTimeInterval:1];
        
        for (ARService *obj in [[ARDiscovery sharedInstance] getCurrentListOfDevicesServices])
        {
            if ([obj.service isKindOfClass:[ARBLEService class]])
            {
                ARBLEService *serviceIdx = (ARBLEService *)obj.service;
                NSLog(@"%@", serviceIdx.peripheral.name);
                //NSString *NAME = @"Delos_DF";
                //NSString *NAME = @"Mykonos_BLE";
                //NSString *NAME = @"Mykonos_DF";
                //NSString *NAME = @"RS_Rouge";
                //NSString *NAME = @"RS_W000207";
                //NSString *NAME = @"RS_W000159";
                //NSString *NAME = @"RS_R000387";
                NSString *NAME = @"RS_B000272";
                //NSString *NAME = @"RS_W000444";
                //NSString *NAME = @"RS_B000497";
                //NSString *NAME = @"RS_B000443";
                //NSString *NAME = @"RS_R000387";
                if ([serviceIdx.peripheral.name isEqualToString:NAME])
                {
                    NSLog(@"%@", serviceIdx.peripheral);
                    foundService = serviceIdx;
                    break;
                }
            }
        }
    }
    
    //[[ARDiscovery sharedInstance] stop];
    ARSAL_Sem_t cancelSem;
    eARSAL_ERROR resultSal= ARSAL_OK;
    ARSAL_CentralManager *centralManager = foundService.centralManager;
    CBPeripheral *peripheral = foundService.peripheral;
    //BOOL ret = NO;
    
    ARSAL_Sem_Init(&cancelSem, 0, 0);
    
    //ARUtils_BLEFtp *bleFtp = [[ARUtils_BLEFtp alloc] init];
    
    [NSThread sleepForTimeInterval:1];
    
    resultSal = [SINGLETON_FOR_CLASS(ARSAL_BLEManager) connectToPeripheral:foundService.peripheral withCentralManager:centralManager];
    
    if  (resultSal == ARSAL_OK)
    {
        resultSal = [SINGLETON_FOR_CLASS(ARSAL_BLEManager) discoverNetworkServices:nil];
    }
    
    //CBService *service = [[_peripheral services]
    /*for (CBService *service in [peripheral services])
    {
        NSLog(@"Service : %@, %@, %04x", [service.UUID representativeString], [service.UUID shortUUID], (unsigned int)service.UUID);
        if([[service.UUID shortUUID] hasPrefix:@"f"])
        {
            resultSal = [SINGLETON_FOR_CLASS(ARSAL_BLEManager) discoverNetworkCharacteristics:nil forService:service];
            
            for (CBCharacteristic *characteristic in [service characteristics])
            {
                NSLog(@"Characteristic : %@, %@, %04x", [characteristic.UUID representativeString], [characteristic.UUID shortUUID], (unsigned int)characteristic.UUID);
            }
        }
    }*/
    for(int i = 0 ; (i < [[peripheral services] count]) && (resultSal == ARSAL_OK) ; i++)
    {
        CBService *service = [[peripheral services] objectAtIndex:i];
        NSLog(@"Service : %@, %@, %04x", [service.UUID representativeString], [service.UUID shortUUID], (unsigned int)service.UUID);
        
        if([[service.UUID shortUUID] hasPrefix:@"f"])
        {
            resultSal = [SINGLETON_FOR_CLASS(ARSAL_BLEManager) discoverNetworkCharacteristics:nil forService:service];
            
            for(int j = 0 ; (j < [[service characteristics] count]) && (resultSal == ARSAL_OK) ; j++)
            {
                CBCharacteristic *characteristic = [[service characteristics] objectAtIndex:j];
                NSLog(@"Characteristic : %@, %@, %04x", [characteristic.UUID representativeString], [characteristic.UUID shortUUID], (unsigned int)characteristic.UUID);
                
                if ([[characteristic.UUID shortUUID] hasPrefix:@"fd23"]
                    || [[characteristic.UUID shortUUID] hasPrefix:@"fd53"])
                {
                    resultSal = [SINGLETON_FOR_CLASS(ARSAL_BLEManager) setNotificationCharacteristic:characteristic];
                    NSLog(@"==REGISTERED Characteristic : %@, %@, %04x", [characteristic.UUID representativeString], [characteristic.UUID shortUUID], (unsigned int)characteristic.UUID);
                }
            }
        }
    }

    if  (resultSal == ARSAL_OK)
    {
        //[bleFtp initWithManager:(__bridge ARNETWORKAL_BLEDeviceManager_t)centralManager device:(__bridge ARNETWORKAL_BLEDevice_t)peripheral];
        //[bleFtp initWithManager:SINGLETON_FOR_CLASS(ARSAL_BLEManager) centralManager:centralManager peripheral:peripheral delegate:nil obj:self];
        //ARUtils_BLEFtp *bleFtp = [[ARUtils_BLEFtp alloc] initWithPeripheral:peripheral cancelSem:NULL port:51];
        //ARUtils_BLEFtp *bleFtp = [[ARUtils_BLEFtp alloc] initWithPeripheral:peripheral cancelSem:&cancelSem port:21];
        ARUtils_BLEFtp *bleFtp = SINGLETON_FOR_CLASS(ARUtils_BLEFtp);
        eARUTILS_ERROR result = ARUTILS_OK;
        
        result = [bleFtp registerPeripheral:peripheral cancelSem:&cancelSem port:21];
        result = [bleFtp registerCharacteristics];
        
        /*if (result == ARUTILS_OK)
        {
            NSMutableString *list = [[NSMutableString alloc] init];
            result = [bleFtp listFiles:@"/update/" list:list];
            
            NSLog(@"LIST: %@", list);
        }*/
        
        if (result == ARUTILS_OK)
        {
            //NSMutableString *list = [[NSMutableString alloc] init];
            char *resultList = NULL;
            uint32_t resultListLen = 0;
            result = [bleFtp listFiles:@"/internal_000/Rolling_Spider/media" resultList:&resultList resultListLen:&resultListLen];
            //result = [bleFtp listFiles:@"/internal_000/Rolling_Spider/thumb" resultList:&resultList resultListLen:&resultListLen];
            //result = [bleFtp listFiles:@"/internal_000/Rolling_Spider/" resultList:&resultList resultListLen:&resultListLen];
            //result = [bleFtp listFiles:@"/Rolling_Spider" resultList:&resultList resultListLen:&resultListLen];
            //result = [bleFtp listFiles:@"/" resultList:&resultList resultListLen:&resultListLen];
            
            NSLog(@"LIST: %s", resultList);
            
            //result = [bleFtp listFiles:@"/internal_000/Rolling_Spider/thumb" resultList:&resultList resultListLen:&resultListLen cancelSem:&cancelSem];
            
            NSLog(@"LIST: %s", resultList);
        }
        
        if (result == ARUTILS_OK)
        {
            NSString *localFile = [NSString stringWithFormat:@"%@/test.jpg", doc];
            
            ARSAL_Sem_Post(&cancelSem);
            
            result = [bleFtp getFile:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-07-08T175119+0000_7D677C6862E81204DF5568597FCC38C3.jpg" localFile:localFile progressCallback:NULL progressArg:NULL cancelSem:&cancelSem];
            
            ARSAL_Sem_Trywait(&cancelSem);
            
            result = [bleFtp getFile:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-07-08T175119+0000_7D677C6862E81204DF5568597FCC38C3.jpg" localFile:localFile progressCallback:NULL progressArg:NULL cancelSem:&cancelSem];
        }
        
        /*if (result == ARUTILS_OK)
        {
            //result = [bleFtp renameFile:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-01-21T160101+0100_3902B87F947BE865A9D137CFA63492B8.jpg" newNamePath:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-01-21T160101+0100_3.jpg"];
            result = [bleFtp renameFile:@"/internal_000/Rolling_Spider/media/test12.jpg" newNamePath:@"/internal_000/Rolling_Spider/media/test13.jpg"];
        }*/
        
        /*if (result == ARUTILS_OK)
        {
            //NSMutableString *list = [[NSMutableString alloc] init];
            char *resultList = NULL;
            uint32_t resultListLen = 0;
            result = [bleFtp listFiles:@"/internal_000/Rolling_Spider/media" resultList:&resultList resultListLen:&resultListLen];

            NSLog(@"LIST: %s", resultList);
        }*/
        
        /*if (result == ARUTILS_OK)
        {
            NSString *localFile = [NSString stringWithFormat:@"%@/test.jpg", doc];
            
            //result = [bleFtp getFile:@"/update/test.jpg" localFile:localFile];
            //result = [bleFtp getFile:@"/update/test.txt" localFile:localFile];
            //result = [bleFtp getFile:@"/update/a.txt" localFile:localFile];
            //result = [bleFtp getFile:@"/update/program.plf" localFile:localFile progressCallback:NULL progressArg:NULL];
            
            //ARSAL_Sem_Post(&cancelSem);
            //result = [bleFtp cancelFile];
            
            //result = [bleFtp getFile:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-01-21T160101+0100_3902B87F947BE865A9D137CFA63492B8.jpg" localFile:localFile progressCallback:NULL progressArg:NULL];
            result = [bleFtp getFile:@"/internal_000/Rolling_Spider/media/Rolling_Spider_2014-06-30T164211+0000_8786A2D82757458FA26DC8127370E5E2.jpg" localFile:localFile progressCallback:NULL progressArg:NULL];
            //result = [bleFtp getFile:@"/internal_000/Rolling_Spider/media/test1.jpg" localFile:localFile progressCallback:NULL progressArg:NULL];
        }*/
        
        /*if (result == ARUTILS_OK)
        {
            uint8_t *data = NULL;
            uint32_t dataLen = 0;
            
            result = [bleFtp getFileWithBuffer:@"/internal_000/Rolling_Spider/thumb/Rolling_Spider_2014-01-21T160101+0100_3902B87F947BE865A9D137CFA63492B8.jpg" data:&data dataLen:&dataLen progressCallback:NULL progressArg:NULL];
        }*/
        
        /*if (result == ARUTILS_OK)
        {
            result = [bleFtp renameFile:@"/a.txt" newNamePath:@"/b.txt"];
        }*/
        
        /*if (result == ARUTILS_OK)
        {
            result = [bleFtp abortPutFile:@"/update/program.plf"];
            if (ret == YES)
            {
                result = [bleFtp deleteFile:@"/update/program.plf"];
            }
        }*/
        
        if (result == ARUTILS_OK)
        {
            NSString *localFile = [NSString stringWithFormat:@"%@/test2.txt", doc];
            
            FILE* src = fopen([localFile UTF8String], "wb");
            //fwrite("0123456789", 1, 10, src);
            char block[132];
            memset(block, '0', sizeof(block));
            //for (int i=0; i<((500 * 10) + 2); i++)
            //for (int i=0; i<((500 * 152)); i++)
            for (int i=0; i<((500 * 2)); i++)
            //for (int i=0; i<((500 * 1)); i++)
            {
                fwrite(block, 1, sizeof(block), src);
            }
            //fwrite(block, 1, 52, src);
            fflush(src);
            fclose(src);
            
            //result = [bleFtp putFile:@"program.plf.tmp" localFile:localFile progressCallback:NULL progressArg:NULL resume:YES cancelSem:&cancelSem];
        }
        
        if (result == ARUTILS_OK)
        {
            //result = [bleFtp deleteFile:@"/update/test2.txt"];
            result = [bleFtp deleteFile:@"program.plf.tmp"];
        }
        
        result = [bleFtp unregisterCharacteristics];
    }
    
    [SINGLETON_FOR_CLASS(ARSAL_BLEManager) disconnectPeripheral:peripheral withCentralManager:centralManager];
    
    [[ARDiscovery sharedInstance] stop];
    
    ARSAL_Sem_Destroy(&cancelSem);
}

@end

File: packages/libARUpdater/TestBench/Linux/autoTest.c
/*
    Copyright (C) 2014 Parrot SA

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in
      the documentation and/or other materials provided with the 
      distribution.
    * Neither the name of Parrot nor the names
      of its contributors may be used to endorse or promote products
      derived from this software without specific prior written
      permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
    OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
    AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.
*/
/**
 * @file autoTest.c
 * @brief libARUpdater TestBench automatic
 * @date 05/23/2014
 * @author djava.bertrand@parrot.com
 */

/*****************************************
 *
 *             include file :
 *
 *****************************************/
#include <stdio.h>
#include <stdlib.h>
#include <libARUpdater/ARUpdater.h>
#include <libARDiscovery/ARDISCOVERY_Discovery.h>
#include <libARSAL/ARSAL.h>

/* ****************************************
 *
 *             define :
 *
 **************************************** */

#define AUTOTEST_TAG "Autotest"


/* ****************************************
 *
 *           variable declarations :
 *
 **************************************** */


/* ****************************************
 *
 *           function declarations :
 *
 **************************************** */




/*****************************************
 *
 *          implementation :
 *
 *****************************************/

void test_http_progress_callback(void* arg, float percent)
{
    char *message = (char *)arg;
    if ((percent == 5) ||
        (percent == 20) ||
        (percent == 50) ||
        (percent == 80) ||
        (percent == 99))
    {
        printf("%s %f%%\n", message ? message : "null", percent);
    }
}

void test_http_should_download_callback(void* arg, int nbDownload, eARUPDATER_ERROR error)
{
    if (error == ARUPDATER_OK)
    {
        if (nbDownload > 0)
        {
            fprintf(stderr, "I should download %i plf file\n", nbDownload);
        }
        else
        {
            fprintf(stderr, "Plf File is up to date\n");
        }
    }
    else if(error == ARUPDATER_ERROR_DOWNLOADER_PHP_APP_OUT_TO_DATE_ERROR)
    {
        fprintf(stderr, "This version is out of date\n");
    }
}

void test_willDownloadPlfCallback(void* arg, eARDISCOVERY_PRODUCT product, const char *const remotePlfVersion)
{
    fprintf(stderr, "Download will start for product : %i (version : %s)\n",product, remotePlfVersion);
}

void test_http_download_completion_callback(void* arg, eARUPDATER_ERROR error)
{
    if (error == ARUPDATER_OK)
    {
        fprintf(stderr, "Download went well\n");
    }
    else
    {
        fprintf(stderr, "Download error !!\n");
    }
}

void test_ftp_upload_callback(void* arg, eARUPDATER_ERROR error)
{
    if (error == ARDATATRANSFER_OK)
    {
        fprintf(stderr, "Upload went well\n");
    }
    else
    {
        fprintf(stderr, "Upload error !!\n");
    }
}

int main(int argc, char *argv[])
{
    eARUPDATER_ERROR error = ARUPDATER_OK;
    eARSAL_ERROR arsalError = ARSAL_OK;
    ARUPDATER_Manager_t *manager = NULL;
    
    ARSAL_MD5_Manager_t* md5Manager = ARSAL_MD5_Manager_New(&arsalError);
    if (arsalError != ARSAL_OK)
    {
        error = ARUPDATER_ERROR_SYSTEM;
    }
    
    if (error == ARUPDATER_OK)
    {
        ARSAL_MD5_Manager_Init(md5Manager);
        manager = ARUPDATER_Manager_New(&error);
    }
    
    if(error == ARUPDATER_OK)
    {
        error =  ARUPDATER_Downloader_New(manager, "./test", md5Manager, ARUPDATER_DOWNLOADER_ANDROID_PLATFORM, "3.0.1",test_http_should_download_callback, NULL, test_willDownloadPlfCallback, NULL, test_http_progress_callback, "test : ", test_http_download_completion_callback, NULL);
        
        if (error == ARUPDATER_OK)
        {
            error = (eARUPDATER_ERROR)ARUPDATER_Downloader_ThreadRun(manager);
        }
        
        fprintf(stderr, "Download finish, uploading now \n");
        if (error == ARUPDATER_OK)
        {
            eARUTILS_ERROR ftpError = ARUTILS_OK;
            ARUTILS_Manager_t *ftpManager = ARUTILS_Manager_New(&ftpError);
            
            if(ftpError == ARUTILS_OK)
            {
                ftpError = ARUTILS_Manager_InitWifiFtp(ftpManager, "172.20.5.48", 21, "", "");
            }
            
            if(ftpError == ARUTILS_OK)
            {
                error = ARUPDATER_Uploader_New(manager, "./test", NULL, ftpManager, md5Manager, ARDISCOVERY_PRODUCT_MINIDRONE, test_http_progress_callback, "prog : ", test_ftp_upload_callback, NULL);
            }
            else
            {
                error = ARUPDATER_ERROR_UPLOADER_ARUTILS_ERROR;
            }
        }
        
        if (error == ARUPDATER_OK)
        {
            error = (eARUPDATER_ERROR)ARUPDATER_Uploader_ThreadRun(manager);
        }
        
        ARUPDATER_Downloader_Delete(manager);
        ARUPDATER_Uploader_Delete(manager);
        ARUPDATER_Manager_Delete(&manager);
    }
    else if(manager != NULL)
    {
        // if there was an error in creating a new manager
        ARUPDATER_Manager_Delete(&manager);
    }
    
    ARSAL_MD5_Manager_Delete(&md5Manager);
    
    fprintf(stderr, "Sum up : %s\n", ARUPDATER_Error_ToString(error));
    
    return 0;
}
